File: Spa.API/Spa.API/Entities/BaiViet.cs
using System.ComponentModel.DataAnnotations;
using System.ComponentModel.DataAnnotations.Schema;

namespace Spa.API.Entities;

[Table("BaiViet")]
public class BaiViet
{
    [Key]
    public int Id { get; set; }

    [Required, MaxLength(250)]
    public string Slug { get; set; } = null!;

    [Required, MaxLength(250)]
    public string TieuDe { get; set; } = null!;

    [MaxLength(500)]
    public string? MoTaNgan { get; set; }

    // Nội dung dạng text dài / HTML / markdown
    public string? NoiDung { get; set; }

    [MaxLength(100)]
    public string? DanhMuc { get; set; }

    [MaxLength(500)]
    public string? AnhBia { get; set; }

    [MaxLength(500)]
    public string? VideoUrl { get; set; }

    [MaxLength(100)]
    public string? TacGia { get; set; }

    public int? ThoiGianDocPhut { get; set; }

    public bool IsPublished { get; set; }

    public DateTime? NgayDang { get; set; }

    public DateTime CreatedAt { get; set; } = DateTime.Now;

    public DateTime? UpdatedAt { get; set; }

    [MaxLength(128)]
    public string? CreatedBy { get; set; }

    [MaxLength(128)]
    public string? UpdatedBy { get; set; }
}

File: admin-spa/src/api/baiViet.ts
import { api } from "../lib/api";

export type BaiVietPublicRow = {
  id: number;
  slug: string;
  tieuDe: string;
  moTaNgan: string | null;
  danhMuc: string | null;
  anhBia: string | null;
  tacGia: string | null;
  thoiGianDocPhut: number | null;
  ngayDang: string; // ISO
};

export type BaiVietPublicDetail = BaiVietPublicRow & {
  noiDung: string | null;
  videoUrl: string | null;
};

export type BaiVietAdminRow = {
  id: number;
  slug: string;
  tieuDe: string;
  moTaNgan: string | null;
  danhMuc: string | null;
  anhBia: string | null;
  videoUrl: string | null;
  tacGia: string | null;
  thoiGianDocPhut: number | null;
  isPublished: boolean;
  ngayDang: string | null;
  createdAt: string;
  updatedAt: string | null;
};

export type BaiVietAdminDetail = BaiVietAdminRow & {
  noiDung: string | null;
};

export type BaiVietCreateDto = {
  slug: string;
  tieuDe: string;
  moTaNgan?: string | null;
  noiDung?: string | null;
  danhMuc?: string | null;
  anhBia?: string | null;
  videoUrl?: string | null;
  tacGia?: string | null;
  thoiGianDocPhut?: number | null;
  isPublished?: boolean;
  ngayDang?: string | null;
};

export type BaiVietUpdateDto = Omit<BaiVietCreateDto, "slug">;

export const baiVietPublicApi = {
  list: async (params: { q?: string; danhMuc?: string; take?: number } = {}) => {
    const sp = new URLSearchParams();
    if (params.q) sp.set("q", params.q);
    if (params.danhMuc) sp.set("danhMuc", params.danhMuc);
    if (params.take !== undefined) sp.set("take", String(params.take));
    const url = sp.toString() ? `/api/public/blog?${sp}` : "/api/public/blog";
    return (await api.get<BaiVietPublicRow[]>(url)).data;
  },
  getBySlug: async (slug: string) =>
    (await api.get<BaiVietPublicDetail>(`/api/public/blog/${encodeURIComponent(slug)}`)).data,
};

export const baiVietAdminApi = {
  list: async (params: { q?: string; published?: boolean; take?: number } = {}) => {
    const sp = new URLSearchParams();
    if (params.q) sp.set("q", params.q);
    if (params.published !== undefined) sp.set("published", String(params.published));
    if (params.take !== undefined) sp.set("take", String(params.take));
    const url = sp.toString() ? `/api/admin/blog?${sp}` : "/api/admin/blog";
    return (await api.get<BaiVietAdminRow[]>(url)).data;
  },
  get: async (id: number) => (await api.get<BaiVietAdminDetail>(`/api/admin/blog/${id}`)).data,
  create: async (dto: BaiVietCreateDto) => (await api.post("/api/admin/blog", dto)).data,
  update: async (id: number, dto: BaiVietUpdateDto) =>
    (await api.put(`/api/admin/blog/${id}`, dto)).data,
  remove: async (id: number) => (await api.delete(`/api/admin/blog/${id}`)).data,
};
